src/sqlfluff/dialects/dialect_hive.py

```diff
@@ -620,3 +621,83 @@ class MsckTableStatementSegment(BaseSegment):
             optional=True,
         ),
     )
+
+
+@hive_dialect.segment(replace=True)
+class FunctionSegment(BaseSegment):
+    """A scalar or aggregate function.
+
+    Extended version of `ansi` to add support of row typecasting
+    https://prestodb.io/docs/current/language/types.html#row
+    ```
+    cast(row(val1, val2) as row(a integer, b integer))
+    ```
+    """
+
+    type = "function"
+    match_grammar = OneOf(
+        Sequence(
+            # Treat functions which take date parts separately
+            # So those functions parse date parts as DatetimeUnitSegment
+            # rather than identifiers.
+            Sequence(
+                Ref("DatePartFunctionNameSegment"),
+                Bracketed(
+                    Delimited(
+                        Ref("DatetimeUnitSegment"),
+                        Ref(
+                            "FunctionContentsGrammar",
+                            # The brackets might be empty for some functions...
+                            optional=True,
+                            ephemeral_name="FunctionContentsGrammar",
+                        ),
+                    )
+                ),
+            ),
+        ),
+        Sequence(
+            Sequence(
+                "ROW",
+                Bracketed(
+                    Delimited(
+                        Sequence(
+                            Ref("BaseExpressionElementGrammar"),
+                            Ref("DatatypeIdentifierSegment", optional=True),
+                        ),
+                    ),
+                ),
```

Comment:
This does not need to be wrapped in another `sequence(...)`.

Also this is causing some of the rules to fail as there is no `function_name` as they expect in a Function.

I tried using `ephemeral_name` but couldn't get that to work so best I could come up with it using `StringParser("ROW", KeywordSegment, type="function_name"),` which is a bit of a fudge, but seems to work:

```suggestion
            # This unusual syntax is used to cast the Keyword ROW to
            # to the function_name to avoid rule linting exceptions
            StringParser("ROW", KeywordSegment, type="function_name"),
            Bracketed(
                Delimited(
                    Sequence(
                        Ref("BaseExpressionElementGrammar"),
                        Ref("DatatypeIdentifierSegment", optional=True),
                    ),
                ),
            ),
```